src/cpp/src/tokenizer.cpp

```diff
@@ -498,7 +490,62 @@ class Tokenizer::TokenizerImpl {
 
 Tokenizer::Tokenizer(const std::filesystem::path& tokenizer_path, const ov::AnyMap& properties) {
     ScopedVar env_manager(tokenizers_relative_to_genai().string());
-    m_pimpl = std::make_shared<TokenizerImpl>(tokenizer_path, properties);
+    auto core = get_core_singleton();
+
+    OPENVINO_ASSERT(tokenizer_path.extension() != ".xml", "'tokenizer_path' parameter should be a path to a dir not a xml file");
+
+    std::shared_ptr<ov::Model> ov_tokenizer = nullptr;
+    std::shared_ptr<ov::Model> ov_detokenizer = nullptr;
+
+    if (std::filesystem::exists(tokenizer_path / "openvino_tokenizer.xml")) {
+        ov_tokenizer = core.read_model(tokenizer_path / "openvino_tokenizer.xml");
+    }
+    
+    if (std::filesystem::exists(tokenizer_path / "openvino_detokenizer.xml")) {
+        ov_detokenizer = core.read_model(tokenizer_path / "openvino_detokenizer.xml");
+    }
+
+    m_pimpl = std::make_shared<TokenizerImpl>(std::make_pair(ov_tokenizer, ov_detokenizer), properties);
+
+    // If special tokens were not found from IR, try to read them from config.
+    // This will work be triggered only for IRs older than 2024.3.
+    if (m_pimpl->m_pad_token_id == -1 || m_pimpl->m_bos_token_id == -1 || m_pimpl->m_eos_token_id == -1) {
+        m_pimpl->read_config(tokenizer_path);
+        m_pimpl->read_special_tokens_map(tokenizer_path);
+        // Try to read tokenizer_config if some token ids or token str are not defined.
+        m_pimpl->read_tokenizer_config_if_necessary(tokenizer_path);
```

Comment:
why logic from pimpl was moved to its wrapper? 

I mean that all this code should be within `TokenizerImpl()` which accepts filepaths

Reply:
We have several ctors, some with `models_path`, some with tokenizer `model_str`, `weight_tensor` or with both buffers, but there is a same logic with applying passes on ov::Model and so on. I could've reused this in other high level ctors like  

`TokenizerImpl(tokenizer_path, ...) : TokenizerImpl(ov_model, ...)` but getting `tokenizer_path` cannot be done a single line, ov_model should be created inside body. But calling other lower lever ctor inside body lookd hacky with `*this = TokenizerImpl` and so on. So making `TokenizerImpl` accept filepath didn't look beautiful.

Therefore i split this stuff in `Tokenizer` and do model reading from path, buffer, or buffers and call relevant `TokenizerImpl`. But this stuff with  `m_pimpl->read_config` and so on look ugly.

**Now I modified approach a bit.** 
I added `setupTokenizer` to avoid use of `*this = TokenizerImpl()` are not used anymore. Now `TokenizerImpl` can be created from filepath. Please take a look.

Reply:
The current approach is not ideal but at least less ugly that it was at the time you commented.

@ilya-lavrenov Can we merge the current latest approach, and improve it separately?